### PR TITLE
helix: update languages.toml generation (support every option in languages.toml)

### PR DESCRIPTION
**Pull request by Philipp-M**
### Description

Updates the generation of `languages.toml` by supporting every option in it with `programs.helix.languages`. This is a breaking change (which IMHO is justified, as it should be more future-proof).

It closes #3911 as well, as it achieves basically the same (support serializing `grammar` and `use-grammars`) with less code (but a breaking change as mentioned above).

But #2871 is not fixed by this, as this seems to be a little bit more tricky (see discussion in this issue).

Motivation for this is support for multiple language servers (https://github.com/helix-editor/helix/pull/2507)
I've included a comment in the example of this option, which I will remove once a new stable helix version is out (but this will likely take some time, as just a new version came out).

### Checklist

- [ ] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [ ] Added myself and the module files to `.github/CODEOWNERS`.

### CODE REVIEW

**Comment by happysalada** [comment]
hey, I've got one suggestion regarding helix.
Would it make sense to merge the original language.toml configuration with the nix-defined ones ? 
The reason  ask is that if currently you want to specify a new language server for a language, then you have to redefine the whole language options (sources, files...) and add the one change to language servers. It seems it would be amazing if you could just say that you want to override the language-servers option and it took the rest from the original configuration file.
Just a thought.

**Comment by ncfavier** [comment]
It seems like something we could do to make the migration easier is to add an assertion that `cfg.languages ? language`.

**Comment by Philipp-M** [comment]
> Would it make sense to merge the original language.toml configuration with the nix-defined ones ?

This happens already, the merge behavior is described/defined in https://github.com/helix-editor/helix/blob/ff442eecd054e67081158ef2f0eeab0ed9156ca5/helix-loader/src/config.rs#L26

The home-manager version just generates the (user-) TOML files. Actually this comment/commit resulted from a PR that was motivated to support exactly this behavior, previously (a long time ago, before the relevant [PR](https://github.com/helix-editor/helix/pull/3080) was merged) only the `language` array was merged.

> It seems like something we could do to make the migration easier is to add an assertion that `cfg.languages ? language`.

That's a very good point, though I'm asserting now that the option is a set (instead of previously a list).
It will pretty much always have to be a set, we could maybe even restrict this in the type, but an assertion with a helpful message may be better for migration.

File: modules/programs/helix.nix [hunk 1]
```diff
@@ -133,16 +143,27 @@ in {
   };
 
   config = mkIf cfg.enable {
+    assertions = [{
+      assertion = builtins.typeOf cfg.languages == "set";
```
Comment (ncfavier) [inline modules/programs/helix.nix:147]:
```suggestion
      assertion = builtins.isAttrs cfg.languages;
```

**Comment by ncfavier** [comment]
An alternative would be to use something like

```nix
with types; coercedTo (listOf tomlFormat.type)
                      (language: lib.warn "boo" { inherit language; })
                      (addCheck tomlFormat.type builtins.isAttrs)
```

**Comment by Philipp-M** [comment]
Nice nix type-fu, I like it. Better than some trace that doesn't immediately show what's wrong, and keeps working, even when the user hasn't updated it (so he can do it at a later time...), I've updated it accordingly, thanks.

File: tests/modules/programs/helix/example-settings.nix [hunk 2]
```diff
@@ -21,10 +21,22 @@ with lib;
         };
       };
 
-      languages = [{
-        name = "rust";
-        auto-format = false;
-      }];
+      languages = {
+        language-server.typescript-language-server = {
+          # depending on derivation outputs will change the generated settings everytime nixpkgs changes...
```
Comment (ncfavier) [inline tests/modules/programs/helix/example-settings.nix:26]: We use [package stubs](https://github.com/nix-community/home-manager/blob/master/tests/stubs.nix) for this; you can look at the other tests for inspiration

Comment (Philipp-M) [reply tests/modules/programs/helix/example-settings.nix]: Thanks for the tip, fixed it in the latest force-push

**Review by ncfavier — APPROVED** [review]

**Comment by KiruyaMomochi** [comment]
For me this warning message is a little confusing:
```plain
warning: The syntax of programs.helix.languages has changed.
It now generates the whole languages.toml file instead of just the language array in that file.

Use
{ language = <languages list>; }
instead.
```

According to the warning message, I have tried to change my config to
```patch
 programs.helix = {
   enable = true;
-  languages = [
+  language = [
     ...
   ]
 };
```

But it does not work. I don't know I should use `languages.language` until I looked into the code.

Therefore, I suggest to change the warning message to
```patch
 Use
-{ language = <languages list>; }
+programs.helix.languages = { language = <languages list>; }
 instead.
```

**Comment by ncfavier** [comment]
Feel free to open a PR

**Comment by mschwaig** [comment]
I was having some trouble with writing a working config that uses this, so I am sharing my code here for setting up `nixd` as a nix language server in case it is useful to others:

```nix
helix = {
  enable = true;
  package = inputs.helix.packages.${pkgs.system}.default;
  languages = {
    language-server.nixd = {
      command = "${pkgs.nixd}/bin/nixd";
    };
    language = [{
      name = "nix";
      auto-format = false;
      language-servers = [ "nixd" "nil" ];
    }];
  };
};
```
which leads to the following `languages.toml` file that the current helix version from master branch is happy with:
```toml
[[language]]
auto-format = false
language-servers = ["nixd", "nil"]
name = "nix"

[language-server.nixd]
command = "/nix/store/v0809ax8pn1glv0vzg5j2dw4zi828mci-nixd-1.2.2/bin/nixd"
```

**Pull request merged**